x-pack/plugins/aiops/server/routes/queries/fetch_significant_categories.ts

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient } from '@kbn/core/server';
+import type { Logger } from '@kbn/logging';
+import { criticalTableLookup, type Histogram } from '@kbn/ml-chi2test';
+import { type SignificantTerm, SIGNIFICANT_TERM_TYPE } from '@kbn/ml-agg-utils';
+
+import type { Category } from '../../../common/api/log_categorization/types';
+import type { AiopsLogRateAnalysisSchema } from '../../../common/api/log_rate_analysis';
+import { LOG_RATE_ANALYSIS_P_VALUE_THRESHOLD } from '../../../common/constants';
+
+import { fetchCategories } from './fetch_categories';
+import { fetchCategoryCounts } from './fetch_category_counts';
+import { getNormalizedScore } from './get_normalized_score';
+
+const getCategoriesTestData = (categories: Category[]): Histogram[] => {
+  const categoriesBaselineTotalCount = getCategoriesTotalCount(categories);
+  return categories.map((d) => ({
+    key: d.key,
+    doc_count: d.count,
+    percentage: d.count / categoriesBaselineTotalCount,
+  }));
+};
+
+const getCategoriesTotalCount = (categories: Category[]): number =>
+  categories.reduce((p, c) => p + c.count, 0);
+
+export const fetchSignificantCategories = async (
+  esClient: ElasticsearchClient,
+  params: AiopsLogRateAnalysisSchema,
+  fieldNames: string[],
+  logger: Logger,
+  // The default value of 1 means no sampling will be used
+  sampleProbability: number = 1,
+  emitError: (m: string) => void,
+  abortSignal?: AbortSignal
+) => {
+  // To make sure we have the same categories for both baseline and deviation,
+  // we do an initial query that spans across baseline start and deviation end.
+  // We could update this to query the exact baseline AND deviation range, but
+  // wanted to avoid the refactor here and it should be good enough for a start.
+  const categoriesOverall = await fetchCategories(
+    esClient,
+    params,
+    fieldNames,
+    params.baselineMin,
+    params.deviationMax,
+    logger,
+    sampleProbability,
+    emitError,
+    abortSignal
+  );
+
+  if (categoriesOverall.length !== fieldNames.length) return [];
+
+  const significantCategories: SignificantTerm[] = [];
+
+  // Using for...of to allow `await` within the loop.
+  for (const [i, fieldName] of fieldNames.entries()) {
```

Comment:
is this list of fields likely to be large? Could it be looped async? or maybe in batches so not to trouble es.